### PR TITLE
MIDI 44: Review CLIP ideas for MIDI data

### PR DESCRIPTION
**Pull request by JasiekKaczmarczyk**
[Review CLIP ideas for MIDI data](https://sacrebleu.atlassian.net/browse/MIDI-44)
[Embedding clustering evaluation](https://sacrebleu.atlassian.net/browse/MIDI-54)
[Streamlit dashboard to review the CLIP embeddings.](https://sacrebleu.atlassian.net/browse/MIDI-56)
[Beat random guess on validation](https://sacrebleu.atlassian.net/browse/MIDI-53)

I added quantization and dataset modules. I also implemented two encoder architectures one for pitch and the other for velocity and time data. Both are based on transformer architecture (more specifically BERT), but I think transformers are not particularly good for long sequences which we'll be dealing with, so maybe a convolutional architecture with attention layers in between (similar to our Unet) would be better. I could also try to implement this: https://arxiv.org/pdf/2302.10866.pdf

### CODE REVIEW

**Comment by JasiekKaczmarczyk** [comment]
I added prepare_dataset.py script for processing maestro dataset to quantized form. I used bin size of 7 for quantization of velocity, dstart and duration. Length of each sequence is 128.

Here is processed dataset: https://huggingface.co/datasets/JasiekKaczmarczyk/maestro-quantized

**Comment by JasiekKaczmarczyk** [comment]
I implemented training loop and just like before I tried overfitting single batch of data.

Here are results:
Zaznaczenie_006

To run training:
```sh
python train.py --config-name <config>
```

Next I will try to run model on whole dataset.

**Comment by roszcz** [comment]
Nice! Can you check what the memory consumption looks like? How large can you make the batch on your machine?

**Comment by JasiekKaczmarczyk** [comment]
> Nice! Can you check what the memory consumption looks like? How large can you make the batch on your machine?

I depends on the model size, but for the model from  ```config-default.yaml``` I used batch_size of 256 and it takes about 19% of memory.

**Comment by roszcz** [comment]
We should be able estimate the batch size more accurately `batch_size = (gpu memory - model size) / sample size` - I'm wondering what's going to be the upper limit for a single gpu training. I can rent a RTX A6000 machine with 48GB - can you guesstimate if it's going to be enough to compare to what's described in the paper?

**Comment by JasiekKaczmarczyk** [comment]
I tried to run training on full dataset for 10 epochs with batch size 512.

Here are results:
Zaznaczenie_007

The accuracy score is very bad, but the model is learning. From what I can see from overfitted model, accuracy score starts getting pretty good when loss reaches about 1.0. I manged to reach about 5.5 loss during training on whole dataset, so I think the problem is with too small model for this task. Also the dataset might be too small.

Run on wandb: https://wandb.ai/jasiekkaczmarczyk/midi-clip/runs/916ecjl5

File: data/prepare_dataset.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+
+def process_record(piece: ff.MidiPiece, sequence_len: int, quantizer: MidiQuantizer) -> list[dict]:
+    piece_quantized = quantizer.quantize_piece(piece)
+
+    midi_filename = piece_quantized.source["midi_filename"]
+
+    record = []
+
+    for subset in tqdm(piece_quantized.df.rolling(window=sequence_len, step=sequence_len)):
```
Comment (roszcz) [inline data/prepare_dataset.py]:
```suggestion
    for subset in tqdm(piece_quantized.df.rolling(window=sequence_len, step=sequence_len // 5)):
```

I think it's ok to have partially overlapping sequences, and the dataset will grow significantly.

It's definitely worth trying out - we could try decreasing the step for as long as the validation performance doesn't deteriorate

**Comment by roszcz** [comment]
@JasiekKaczmarczyk do you know what's up with those dips in the `train/loss` chart? Is this a result for a smaller batch size at the end of the epoch? If that's the case, would it make sense to normalize the loss with regard to batch size?

File: data/prepare_dataset.py (hunk 1 above)
Comment (JasiekKaczmarczyk) [reply data/prepare_dataset.py]: Yeah, that could work. I'll look into it.

File: data/prepare_dataset.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import os
+
+import fortepyan as ff
+from tqdm import tqdm
+from quantizer import MidiQuantizer
```
Comment (roszcz) [inline data/prepare_dataset.py]:
```suggestion
from data.quantizer import MidiQuantizer
```

File: data/prepare_dataset.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+if __name__ == "__main__":
+    # get huggingface token from environment variables
+    token = os.environ["HUGGINGFACE_TOKEN"]
+
+    # hyperparameters
+    sequence_len = 128
+
+    # That's where I'm downloading the LTAFDB data
+    hf_dataset_path = "roszcz/maestro-v1"
```
Comment (roszcz) [inline data/prepare_dataset.py]:
```suggestion
    hf_dataset_path = "roszcz/maestro-v1-sustain"
```

This is a version of the dataset with the sustain pedal applied. This makes the data much closer to the original performance.

File: data/prepare_dataset.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            "duration_bin": subset.duration_bin.astype("int16").values.T,
+            "velocity_bin": subset.velocity_bin.astype("int16").values.T,
+        }
+
+        record.append(sequence)
+
+    return record
+
+
+if __name__ == "__main__":
```
Comment (roszcz) [inline data/prepare_dataset.py]:
```suggestion
def main():
...



if __name__ == "__main__":
    main()
```

`__main__` shouldn't contain that much logic. The main plus here is that with this change I can run `prepare_dataset.main()` in ipython :)

File: data/prepare_dataset.py (hunk 1 above)
Comment (roszcz) [inline data/prepare_dataset.py]:
```suggestion
    step = sequence_len // 2
    window_split = piece_quantized.df.rolling(window=sequence_len, step=step)
    for subset in tqdm(window_split, leave=False):
```

You can have nested progress bars with `leave=False`

File: train.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    val_ds = MidiDataset(dataset_name, split="validation")
+    test_ds = MidiDataset(dataset_name, split="test")
+
+    if overfit_single_batch:
+        train_ds = Subset(train_ds, indices=range(batch_size))
+        val_ds = Subset(val_ds, indices=range(batch_size))
+        test_ds = Subset(test_ds, indices=range(batch_size))
+
+    # dataloaders
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers)
```
Comment (roszcz) [inline train.py]:
```suggestion
    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers, shuffle=True)
```

I think it's necessary to shuffle the train split - otherwise batches will be coming in creation order - mostly from the same piece and in sequence

File: configs/config-default.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+paths:
+  save_ckpt_dir: "checkpoints" # directory where checkpoints will be saved
+  load_ckpt_path: null # if not None, specifies path to model state dict which will be loaded
+  log_dir: "logs"
+  hf_repo_id: null # repo id to upload model to huggingface if null model is not uploaded
+
+models:
+  pitch_encoder:
+    num_embeddings: 128
```
Comment (roszcz) [inline configs/config-default.yaml:25]: Technically there's only 88 keys on a piano :)

**Comment by JasiekKaczmarczyk** [comment]
I added script to visualize embeddings of model. I used Kmeans to cluster pitch embeddings to 10 clusters and then applied PCA to reduce dimensions and check if clusters are in roughly the same place in velocity/time embedding space as in pitch embedding space.

Here are results on batch from train set:
embeddings_clustering_train_set

Here are results on batch from validation set:
embeddings_clustering_val_set

The model starts getting rough shape of the data distribution, but clearly is not trained yet. Both encoders have about 1 million parameters. My guess to improve results is to increase model size and batch size, so it has more samples to compare during one forward pass.

**Comment by roszcz** [comment]
I think we should also try to check embedding consistency on other properties of the data - are the fragments in a cluster actually similar? For example, are Chopins close to Chopins in the embedding space? Are sonatas close to sonatas?

**Comment by JasiekKaczmarczyk** [comment]
I added new function to ```embedding_clustering.py``` to visualize embeddings based on some query.

Here are some results:
embeddings_query_chopin

embeddings_query_liszt

embeddings_query_verdi

**Comment by roszcz** [comment]
Nice, looks promising 🚀

**Comment by JasiekKaczmarczyk** [comment]
> Nice, looks promising 🚀

I think now main bottleneck is batch size. This model used batch size of 1024 and was trained for 10 epochs, dataset has 239k samples. This training technique relies heavily on seeing as much data in one pass as possible to make meaningful embeddings. Other solution is to train model longer, so each sample has more data for contrasting.

Also I think changing aggregation function for global embedding in encoders from ```torch.mean()``` to something else (maybe CLS token) might also help.

**Comment by roszcz** [comment]
I think we need more ways to verify the quality of those embeddings, and I'd say the best thing to do is to start listening. We don't really have any meaningful way of checking if anything is improving other than looking at the loss value, which I think can be misleading

**Comment by JasiekKaczmarczyk** [comment]
> 

I don't think that is very efficient. I could try implement it but I don't know if it would be informative, especially since velocity, dstart and duration are quantized.

**Comment by JasiekKaczmarczyk** [comment]
I created streamlit dashboard to review CLIP embeddings. You can provide query for pitch embeddings and velocity/time embeddings separately. I used PCA to visualize embeddings. KDTree is used to find two closest embeddings, one in each set. Finally mp3 file is created in tmp directory and loaded into streamlit.

To run it:
```sh
streamlit run --server.port 4466 tests/streamlit_dashboard.py
```

**Comment by roszcz** [comment]
Thanks!

To run this, I have to set up python path manually, e.g.:

```sh
PYTHONPATH=. streamlit run --server.port 4466 tests/streamlit_dashboard.py
```

Can you add support for the Maestro dataset? It is much higher quality than Giant Midi, so we should prioritize it for validation. I think you'll need to update you quantized clone to track the compose and piece title better - I did this in my code review PR, so hopefully it'll be low effort :)

**Comment by JasiekKaczmarczyk** [comment]
> Thanks!
> 
> To run this, I have to set up python path manually, e.g.:
> 
> ```shell
> PYTHONPATH=. streamlit run --server.port 4466 tests/streamlit_dashboard.py
> ```
> 
> Can you add support for the Maestro dataset? It is much higher quality than Giant Midi, so we should prioritize it for validation. I think you'll need to update you quantized clone to track the compose and piece title better - I did this in my code review PR, so hopefully it'll be low effort :)

Ok, I've updated it and used your maestro-quantized with source value for querying. Could you also make your model checkpoint visible on huggingface so I could use it for testing?

**Comment by roszcz** [comment]
> Could you also make your model checkpoint visible on huggingface so I could use it for testing?

Here it is: https://huggingface.co/roszcz/midi-clip/blob/main/midi-clip-batch-256-2023-08-27-10-14.ckpt

And here's how it looks in wandb:

image

**Review by roszcz — APPROVED** [review]
